packages/editor/src/plugins/ListsPlugin/ListsPlugin.tsx

```diff
@@ -357,79 +364,101 @@ export class ListsPlugin implements IPlugin {
   };
 
   static toggleList = (
     editor: Editor,
     listType: string,
     listItemType = ListsPlugin.LIST_ITEM_ELEMENT,
+    defaultType = "paragraph"
   ) => {
+    function isList(node: CustomElement) {
+      return (
+        Element.isElement(node) &&
+        [
+          ListsPlugin.BULLETED_LIST_ELEMENT,
+          ListsPlugin.NUMBERED_LIST_ELEMENT,
+        ].includes(node.type)
+      );
+    }
+
     Editor.withoutNormalizing(editor, () => {
-      const node = Editor.above(editor, {
-        match: (n) =>
-          [
-            ListsPlugin.BULLETED_LIST_ELEMENT,
-            ListsPlugin.NUMBERED_LIST_ELEMENT,
-          ].includes((n as CustomElement).type),
+      const listEntry = Editor.above(editor, {
+        match: (n) => {
+          return isList(n as CustomElement);
+        },
+        mode: "lowest",
       });
 
-      if (node) {
-        Transforms.setNodes(
-          editor,
-          {
-            type: listType,
-          },
-          {
-            match: (n) =>
-              [
-                ListsPlugin.BULLETED_LIST_ELEMENT,
-                ListsPlugin.NUMBERED_LIST_ELEMENT,
-              ].includes((n as CustomElement).type),
-          },
-        );
-      } else {
-        const { selection } = editor;
+      const { selection } = editor;
 
-        if (!selection) {
-          return;
-        }
+      if (listEntry) {
+        if (!selection) return;
 
-        const nodes = Array.from(
-          Editor.nodes(editor, {
-            at: Editor.unhangRange(editor, selection),
-            match: (n) => Element.isElement(n),
-            mode: 'highest',
-          }),
-        );
+        const [node, path] = listEntry;
+        const currentListType = (node as CustomElement).type;
+
+        const range = Editor.unhangRange(editor, selection);
 
-        if (!nodes.length) {
+        if (currentListType === listType) {
+          Transforms.setNodes(
+            editor,
+            { type: defaultType },
+            {
+              at: range,
+              match: (n) => {
+                return Element.isElement(n) && n.type === listItemType;
+              },
+              mode: "highest",
+            }
+          );
+        } else {
+          Transforms.setNodes(
+            editor,
+            { type: listType, children: [] },
+            {
+              at: path,
+              match: (n) => node === n,
+            }
+          );
+        }
+      } else {
+        if (!selection) {
           return;
         }
 
-        Transforms.wrapNodes(
+        const range = Editor.unhangRange(editor, selection);
+
+        Transforms.setNodes(
           editor,
-          { type: listType, children: [] },
+          { type: listItemType },
           {
-            mode: 'highest',
+            at: range,
             match: (n) => {
-              return Element.isElement(n) && n.type === listItemType;
+              return Element.isElement(n) && n.type === defaultType;
```

Comment:
As far as I understood this code, we're converting only defaultType elements to the list. What if I want to convert a heading into the list? I think this code won't work